### PR TITLE
implemented NCA pre-pretraining (val/loss - 3.207276)

### PR DESCRIPTION
**Pull request by idhantgulati**
this PR includes the implementation of NCA pre-pretraining (orginal paper: https://arxiv.org/abs/2603.10055, orginal open-source code: https://github.com/danihyunlee/nca-pre-pretraining)

**final best val/loss:** 3.207276

**total training time:** 71m (pre-pre-training) +  240m (pre-training) = **5h 11m**
**total wall time:** 113m (pre-pre-training) + 272m (pre-training) = 6h 25m

training command / script used for this is present in `run.sh` 

**training stats report:** [wandb report](https://api.wandb.ai/links/moe-research/dybafcio)

## nca pre-pre-training:

implentation for the nca pre-pretraining is present in `pre_pre_train.py`

for this 50M were used per-epoch for 6 epochs (i.e., 300M total NCA tokens) 

```bash
WANDB_MODE=online torchrun --standalone --nproc_per_node=8 pre_pre_train.py \
    --tokens-per-epoch 50000000 --num-eval-tokens 16000000 \
    --num-epochs 6 \
    --regen-data \
    --grid 12 --patch 2 --num-colors 10 \
    --temperature 0.0001 --dT 1 --init-rollout-steps 10 \
    --filter-rules --filter-threshold 0.5 --filter-upper-bound 1.0 \
    --n-layer 30 --n-head 20 --n-embd 2560 \
    --seq-len 1024 --window-pattern SSSL \
    --device-batch-size 8 --total-batch-size 65536 \
    --lr 6e-4 --weight-decay 0.3 --dropout 0.1 \
    --warmdown-ratio 0.15 \
    --save-dir nca_ckpts/ppt_v8 \
    --run nca-ppt-v8
```
final train/loss: 3.339553
final val/loss: 3.420781

log - [nca-ppt-training.log](https://github.com/user-attachments/files/26522804/nca-ppt-training.log)

## pre-training:

also some edits were made to the `train.py` in order to accomodate pre-pre-training like 

```bash
WANDB_MODE=online torchrun --standalone --nproc_per_node=8 train.py \
    --run lang-pre-train-v10 \
    --pretrained-ckpt nca_ckpts/ppt_v8/nca_pretrained_best.pt \
    --n_embd 2560 --n_head 20 --n_layer 30 \
    --num-epochs 25 \
    --total-batch-size 524288 \
    --lr_multiplier 0.25 \
    --weight-decay 1.3 \
    --dropout 0.1 \
    --nca-load-mode attn \
    --nca-warmup-steps 250 --nca-rampup-steps 200 \
    --warmdown-ratio 0.2 \
    --dupe-start-epoch 12 \
    --logit-avg 5 \
    --swa-last-epochs 4
```
log - [pre-training.log](https://github.com/user-attachments/files/26522814/pre-training.log)

### compute
all the training was done on a single node of 8xH100 (via modal.com)

> PS: noticed that during the end of training the model started to overfit due to which the val/loss started to increase. hence, the last couple of epochs were not included in the final best val/loss.

> also thanks to @MarmikChaudhari for the inital idea :)

### CODE REVIEW

**Comment by idhantgulati** [comment]
thats sounds fair. i can run the training again in order to get it comparable against the leaderboard. 

just to get it clear, will be running a training run on 1.4B model. are there any other params to match the i keep have. would. it be fine to have it run for longer epochs / time. 

so i'll be running 2 training runs (1) w/ NCA and (2) w/o NCA. does that sound good?

**Comment by idhantgulati** [comment]
@akshayvegesna

**Comment by akshayvegesna** [comment]
> so i'll be running 2 training runs (1) w/ NCA and (2) w/o NCA. does that sound good?

Yes that sounds good! If there is a gain from NCA, this is good to go. Any parameters that you use in the Fineweb training with NCA should be match the Fineweb training configuration without NCA (including epoch count).

**Comment by idhantgulati** [comment]
sounds good! 

i'll just the same training run that i did before. with just the right model size.

**Comment by samacqua** [comment]
When the paper came out, I tried running the NCA experiments using their repo but with 2 slight differences: a). using Muon on the 2d weights and b). using Fineweb as the dataset. The NCA gains held when using Adam and Fineweb. But, when I used Muon, NCA actually *hurt* the performance -- making it comparable to Adam w/ no NCA (step-wise comparison).

So roughly, Muon (NCA) ~= Adam (no NCA), and Muon (no NCA) ~= Adam (NCA).

My rough intuition for why is that the spectral normalization of Muon makes the weight updates "move more" than in Adam, kind of like how Muon erases the number encoded in the weights like below (from https://jeremybernste.in/writing/deriving-muon). Given this, I tried a more gentle LR warmup when init'ing after NCA pre-pre-training, but that didn't change results much.

image


Given that NCA is a fixed cost and Muon is per-step overhead, maybe there is some step count where Adam+NCA does better than Muon w/ no NCA when measuring wall-clock time. However, I don't think this is the case, but it would be interesting for someone to explore this more fully.

Given the many implementation choices (e.g. maybe mixing the pre-pre-training data throughout pre-training, different Muon requires different NCA hparams, ...), a negative result here doesn't mean **that** much, but putting this in case it is useful. I love the idea of algorithmic pre-pre-training and hope someone can get it to work :)

**Comment by idhantgulati** [comment]
@samacqua this is very interesting observation and helpful in this case. thank you. 

also i had a question, did you keep the optimizer constant (like muon / adam) for both pre-pre-training and pre-training phase? 

further, what about the other hyperparams like LR, WD, etc. did you change / tune them? if so, to what degree?

**Comment by idhantgulati** [comment]
so couple of training runs with sligthly different configs. with the model paramaters being - `1,398,287,931 (transformer: 1,169,829,360, value_embeds: 48,168,960, lm_head: 90,144,768, other: 90,144,843)`

didnt really that much sucess interms of val/loss. ran each of the runs for 25ep while changing some of the NCA specific params 

final min val/loss 
baseline; **w/o NCA** - 3.177205 [2 hours 6 min]
it-1 - 3.187808 
it-2 - 3.208140 [bumped up ppt epochs from 6 -> 10; also used nca load mode = all instead of attn; bumped warm/ramp = 500/500 from 250/200]
it-3 - 3.205975 [nca load mode = all; w/ warmup steps = 0, rampup = 500]
it-4 - 3.193332 [back to load mode attn; w/ warm/ramp = 500/500]

after looking into what @samacqua. i have an idea or two to try in mind to try. i was thinking to try hybrid optimizer approach wherein adamW only on the matrix params that were loaded from the NCA checkpoint; muon on everything else. 

but apart from what if theres any other feedback or so. open to that.

**Comment by samacqua** [comment]
> also i had a question, did you keep the optimizer constant (like muon / adam) for both pre-pre-training and pre-training phase?

Yes, I tried the cross product of `(no pre-pre-training, adam pre-pre-training, muon pre-pre-training)` and `(adam pretraining, muon pretraining)`. I just found the old results:

image

Barely tuned hparams. Only change was to increase the LR warmup or muon pretraining in the hopes that this would preserve the pre-pre-training structure, to no avail. So muon pre-pre-training gets better pre-pre-training loss *and* helps downstream pre-training more for both pre-training optimizers.
